codestyle/styles.md

```diff
@@ -0,0 +1,672 @@
+- Will it be comfortable for you?
+- Does it work for your team?
+
+If you found a grammar mistake or would like to reformulate
+something for better understanding let me know as well.
+
+---
+
+<a id="base"></a>
+## Base requirements
+
+* UTF-8 encoding without BOM;
+* unix-style line endings;
+* 80-characters line wide where possible;
+* four (4) spaces indents, no tabs;
+* no trailing spaces (empty lines must contain only the line feed symbol);
+* lowercase only, except custom values in quotes, url's and HEX-values;
+* single quotes;
+* one (1) empty line at the end of files;
+
+
+
+<a id="ruleset"></a>
+## Ruleset
+
+The standard ruleset looks like:
+
+```CSS
+[selector] {
+    [property]: [value];
+    [<—declaration—>]
+}
+```
+
+for example:
+
+```CSS
+.block {
+    display: block;
+    background-color: green;
+    color: red;
+}
+```
+
+* each ruleset starts on its own new line;
+* the opening brace (`{`) on the same line as the last selector;
+* one (1) space between the last selector and the opening brace (`{`);
+* each declaration on its own new line indented by four (4) spaces;
+* no empty lines between declarations, except grouping properties;
+* properties and values on the same line;
+* no spaces between property and colon (`:`);
+* one (1) space between colon (`:`) and value;
+* each declaration ends with a semicolon (`;`);
+* no spaces between the value and the semicolon (`;`);
+* the closing brace (`}`) on its own new line, indented as the last
+  selector;
+
+**Grouping selectors**
+
+Two or more selectors have to be separated by comma:
+
+```CSS
+.foo,
+.bar,
+.baz {
+    display: block;
+    color: red;
+}
+```
+
+* each selector on its own new line;
+* no empty lines between grouped selectors;
+* each selector in the list except the last one ends with a comma (`,`);
+* no spaces between selector and comma (`,`);
+
+**Grouping properties**
+
+You can group related properties adding one (1) empty line between groups of
+properties:
+
+```CSS
+.block {
+    position: absolute;
+
+    border-style: solid;
+    border-width: 1px 0 0 1px;
+    border-color: #FFF;
+
+    background-color: #000;
+    background-position: left top;
+}
+```
+
+This record will be wrong because grouped properties are not related to each
+other:
+
+```CSS
+.block {
+    display: block;
+
+    font-size: 30px;
+    background: #EEE;
+}
+```
+
+**Empty rulesets**
+
+Empty rulesets are not allowed. A ruleset is considered as empty if it doesn't
+contain at least one styling rule. These record will be wrong:
+
+```CSS
+.block {}
+
+.comment-inside {
+    /*
+        Still empty because comments are not the styling rules
+    */
+}
+```
+
+The only exception is if a block doesn't have any styling rules, but its
+elements have you can keep the block ruleset empty:
+
+```CSS
+.block {}
+
+    .block__element1 {
+        background: #F00;
+    }
+
+    .block__element2 {
+        background: #0F0;
+    }
+```
+
+**Indenting rulesets**
```

Comment:
I’ve meet this thoughts before and it sounds ideologically right. Opponents the idea mostly express two main concerns.
The first one is that BEM doesn’t suppose to have any structure inside the block. This is how it was designed to work: we have a  root block element and many elements inside it, no matter how deep in the block they are placed. The second concern is that the structure of the block may be changed or elements inside the block can be used in different parts interchangeably. Support the actual structure of the block in such cases in the CSS-file is impossible.

I personally worked over two years adhering the practice to place all elements in the plain list without any structure and despite the fact it sounds so ideologically right I came to realize that it’s completely not comfortable approach.

The biggest problem here lies in how we work with code, especially how we read it. Since working with huge arrays of uniform information is hard for the brain, the human eye is always trying to seek some visual anchors in the text to help the brain find desired information without parsing all the array of data. Indents help us to read and understand code much easier.

Just open the links in two tabs and compare what is easier to read:
[BEM block WITHOUT indents](https://gist.github.com/deemsk/8cbe676a7b619c14b236)
[BEM block WITH indents](https://gist.github.com/deemsk/56d62b5634d53637aea8)

Here is what Harry Roberts write in his CSS Guidelines about the subject: “As well as indenting individual declarations, indent entire related rulesets to signal their relation to one another:
```
.foo {}

    .foo__bar {}

        .foo__baz {}
```
By doing this, a developer can see at a glance that `.foo__baz {}` lives inside `.foo__bar {}` lives inside `.foo {}`. This quasi-replication of the DOM tells developers a lot about where classes are expected to be used without them having to refer to a snippet of HTML.” http://cssguidelin.es

Also I don’t share the concern about blocks with a flexible structure. In fact I noticed that there are very few cases when it can occurs. In such rare cases we can place the ruleset in any part of the block where it can be used. I believe that we shouldn’t avoid the practice which help us in 99% of cases and could be doubtful in 1%.